src/_posts/2020-04-28-terraform-with-multiple-environments.md

```diff
@@ -0,0 +1,167 @@
+
+As an example, `env.dev.tfvars` could include
+
+```
+retention_policy = 7
+```
+
+and `prod.env.tfvars` would include
+
+```
+retention_policy = 365
+```
+
+The other option, which is the one I used, is to create maps that have the necessary values, and then using the lookup functionality to get the information, so the code on your **tf** file could look like this:
+
+```
+variable "workspace_to_retention_policy_map" {
+  description = "The retention period of cloudwatch logs"
+  type = map
+  default = {
+    dev     = 7
+    staging = 7
+    prod    = 365
+  }
+}
+
+locals {
+  retention_policy  = lookup(var.workspace_to_retention_policy_map, terraform.workspace, 7)
+  # or even better if you don't need the default
+  retention_policy2 = var.workspace_to_retention_policy_map[terraform.workspace]
+}
+```
+
+I actually like having together the values that change per environment, so it easy to see what changes between environment, but it does create far more noise within the file.
+
+As I said, at that point workspaces were not recommended for multiple environments. You have to be quite aware on when to do `terraform workspace select`, or better only do it through your ci, so you don't make any mistake.
+
+# The second way - Multiple accounts, single workspace
+
+At my current project we are using a different approach. We have an AWS account per environment. We could have used the same workspace setup, but that is complicated by the fact that the state is stored on the AWS account using S3 (solution, next section).
+
+So the approach was to have an `environments` folder, and inside a folder per environment. Each folder (dev, test, prod, whatever you use) has the same files (your variables file, your outputs file, your main/setup files). But as you should have all environments looking the same (or pretty similar), what you can do is symlink those files, and add a tfvars file for the data that needs to change (not disimilar to what we saw on [The first way](#thefirstway-singleaccountwithworkspaces)). Any infrastructure that is only present on an environment could be added into an additional file(s). You will have something like this
```

Comment:
```suggestion
So the approach was to have an `environments` folder, and inside a folder per environment (dev, test, prod, whatever you use). Each folder has the same files (your variables file, your outputs file, your main/setup files). Because you should have all environments looking the same (or pretty similar), what you can do is symlink those files, and add a **tfvars** file for the data that needs to change (not disimilar to what we saw on [The first way](#thefirstway-singleaccountwithworkspaces)). Any infrastructure that is only present on an environment could be added into an additional file(s). You will have something like this
```